src/executor/aot.rs

```diff
@@ -0,0 +1,367 @@
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+{
+    library: Library,
+    registry: ProgramRegistry<TType, TLibfunc>,
+}
+
+impl<TType, TLibfunc> AotNativeExecutor<TType, TLibfunc>
+where
+    TType: GenericType,
+    TLibfunc: GenericLibfunc,
+    <TType as GenericType>::Concrete: TypeBuilder<TType, TLibfunc> + ValueBuilder,
+{
+    pub fn new(library: Library, registry: ProgramRegistry<TType, TLibfunc>) -> Self {
+        Self { library, registry }
+    }
+}
+
+impl AotNativeExecutor<CoreType, CoreLibfunc> {
+    pub fn invoke_dynamic(&self, function_id: &FunctionId, args: &[JitValue]) -> JitValue {
+        let function_name = generate_function_name(function_id);
+        let function_name = format!("_mlir_ciface_{function_name}");
+
+        // Arguments and return values are hardcoded since they'll be handled by the trampoline.
+        let function_ptr = unsafe {
+            self.library
+                .get::<extern "C" fn()>(function_name.as_bytes())
+                .unwrap()
+        };
+
+        let function_signature = &self.registry.get_function(function_id).unwrap().signature;
+        let mut invoke_data = Vec::new();
+
+        // Generate return pointer (if necessary).
+        let return_ptr = if function_signature.ret_types.len() > 1
+            || function_signature
+                .ret_types
+                .last()
+                .is_some_and(|id| self.registry.get_type(id).unwrap().is_complex())
+        {
+            let layout =
+                function_signature
+                    .ret_types
+                    .iter()
+                    .fold(Layout::new::<()>(), |layout, id| {
+                        let type_info = self.registry.get_type(id).unwrap();
+                        layout
+                            .extend(type_info.layout(&self.registry).unwrap())
+                            .unwrap()
+                            .0
+                    });
+
+            let return_ptr = unsafe { std::alloc::alloc(layout) };
+            invoke_data.push(return_ptr as u64);
+
+            Some((layout, return_ptr))
+        } else {
+            invoke_data.push(0);
+            None
+        };
+
+        for (type_id, value) in function_signature.param_types.iter().zip(args) {
+            map_arg_to_values(&mut invoke_data, &self.registry, type_id, value).unwrap();
+        }
+
+        // Invoke the trampoline.
+        let mut ret_registers = [0; 4];
+        unsafe {
+            aot_trampoline(
+                function_ptr.into_raw().into_raw(),
+                invoke_data.as_ptr(),
+                invoke_data.len(),
+                &mut ret_registers,
+            );
+        }
+
+        // Parse return values.
+        let type_info = self
+            .registry
+            .get_type(function_signature.ret_types.last().unwrap())
+            .unwrap();
+        let return_value = match type_info {
+            CoreTypeConcrete::Array(_) => match return_ptr {
```

Comment:
I think you also need to handle the snapshot return type, which is basically the inner type. On the from_jit we do this

`CoreTypeConcrete::Snapshot(info) => Self::from_jit(ptr, &info.ty, registry),`

but i think you should also add this case here. Afaik snapshots can only be made for arrays (maybe structs but i havent seen it)